### PR TITLE
[KEP-2395] Phase 4 - Disabling In-Tree Providers

### PR DESCRIPTION
**Pull request by dims**
KEP-2395 states that Beta was targeted for v1.26, So we missed the boat and now it's ~v1.28~ v1.29 cycle :) Let's see if we can land this now and adjust the CI jobs as needed.

https://github.com/kubernetes/enhancements/tree/master/keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers#phase-4---disabling-in-tree-providers

DisableCloudProviders - this feature gate will disable any functionality in kube-apiserver, kube-controller-manager and kubelet related to the --cloud-provider component flag.

DisableKubeletCloudCredentialProvider - this feature gate will disable in-tree functionality in the kubelet to authenticate to the Azure and GCP container registries for image pull credentials.

KEP metadata update in https://github.com/kubernetes/enhancements/pull/4171/files

#### What type of PR is this?
/kind feature

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
In tree cloud providers are now switched off by default. Please use DisableCloudProviders and DisableKubeletCloudCredentialProvider feature flags if you still need this functionality.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by BenTheElder** [comment]
There's "unit tests" for these scripts that will need tweaking

So this generally means CI will not be covering them now?

**Comment by dims** [comment]
@BenTheElder that's what i am trying to figure out here, which tests we will fail and will need another place/way to test. so far it seems to be just `TestMetadata`

**Comment by BenTheElder** [comment]
> @BenTheElder that's what i am trying to figure out here, which tests we will fail and will need another place/way to test. so far it seems to be just TestMetadataClient

Sorry, to clarify: I mean right now we defacto cover what's left of these in ~all CI jobs? But we will not after this PR (and we need to make sure the out-of-tree provider is running?), or did we already switch and we're just turning off the feature gate and expect most jobs to be untouched?

Need to catch up on this, we're hoping to get more external cloud provider codepath coverage with https://github.com/kubernetes-sigs/cloud-provider-kind soon.

**Comment by aojea** [comment]
Did we finish phase 3 https://github.com/kubernetes/enhancements/tree/master/keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers#phase-3---migrating-provider-code-to-provider-repos  

/assign @andrewsykim

**Comment by dims** [comment]
> Did we finish phase 3

Azure, vSphere, GCP all have working out-of-tree providers.

**Comment by CecileRobertMichon** [comment]
@andyzhangx @dims the azure test failures are due to the fact the presubmit jobs are still using in-tree cloud-provider (everything else is testing out of tree but we kept the PR jobs to in-tree to be able to test any potential critical bug fixes in the in-tree cloud-provider):

```
Apr 21 09:09:10.677959 capz-ekmb5k-control-plane-fr2kj kubelet[2424]: E0421 09:09:10.677866    2424 run.go:74] "command failed" err="failed to run Kubelet: failed to create kubelet: cloud provider \"azure\" was specified, but built-in cloud providers are disabled. Please set --cloud-provider=external and migrate to an external cloud provider"
```

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/117503/pull-kubernetes-e2e-capz-azure-disk/1649335795573592064/artifacts/clusters/capz-ekmb5k/machines/capz-ekmb5k-control-plane-h5llb/kubelet.log

I actually have a PR in flight to switch them to out-of-tree (convenient timing!) so once that merges we should be able to rerun the capz/azure tests here using external cloud-provider-azure.

**Comment by dims** [comment]
> so once that merges we should be able to rerun the capz/azure tests here using external cloud-provider-azure.

Sweeeeeettt! thanks @CecileRobertMichon

**Comment by CecileRobertMichon** [comment]
> @CecileRobertMichon what is missing to merge it?

I'm still testing it locally and running into some errors, hoping to have those resolved by EOD. Then I'll mark the PR ready for review once it's approved we can merge it and backport it to the CAPZ release-1.8 branch since that's what the jobs above are using.

**Comment by CecileRobertMichon** [comment]
Update https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3456 is now ready for review

**Comment by CecileRobertMichon** [comment]
All 4 Azure tests are now passing and running with out of tree @aojea @dims

**Comment by aojea** [comment]
Andrew fixed the job too, so we should be good to go


/test pull-e2e-gce-cloud-provider-disabled

**Review by andrewsykim — COMMENTED** [review]
Thanks for starting this PR! I added this for discussion in the next SIG meeting.

(holding til then)

/hold

File: pkg/features/kube_features.go
Comment (andrewsykim) [inline pkg/features/kube_features.go:984]: Also update the feature gate comments please

Comment (dims) [reply pkg/features/kube_features.go:984]: Done!

**Review by elmiko — COMMENTED** [review]
this is mostly making sense to me. i like removing the feature gate transformation from the config-default.sh, that seems less brittle. is the idea here to get all feature configuration from the test config?

**Comment by dims** [comment]
> is the idea here to get all feature configuration from the test config?

Yep looking at the test definitions should reflect what we are testing (both the external provider and feature gate combinations). And we enforce the valid combinations in the code as well (to prevent people tripping over setting one thing and not the other). We can relax the hard failure to a warning/error if we find it necessary as we go on in 1.29 release cycle.

**Comment by elmiko** [comment]
i'm happy with the direction this is going, but i'd like get another review and see the tests passing. i've talked with @dims on slack and he is going find another reviewer who can take a look on monday. _fyi monday is a holiday in the usa_

/hold cancel

**Comment by dims** [comment]
@aojea please look at failures in the worker node kubelet in the latest `pull-kubernetes-e2e-gci-gce-ingress` CI job:

from log - https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/117503/pull-kubernetes-e2e-gci-gce-ingress/1697765618037559296/artifacts/bootstrap-e2e-minion-group-15cv/kubelet.log:
```
Sep 02 00:46:52.046603 bootstrap-e2e-minion-group-15cv kubelet[10550]: E0902 00:46:52.046567   10550 run.go:74] "command failed" err="failed to run Kubelet: cannot create certificate signing request: certificatesigningrequests.certificates.k8s.io is forbidden: User \"kubelet\" cannot create resource \"certificatesigningrequests\" in API group \"certificates.k8s.io\" at the cluster scope"
```

**Comment by dims** [comment]
/assign @aojea @upodroid @tzneal 

Let's get this in ... in the weekend itself, so we can deal with other failures which may occur.

File: pkg/features/kube_features.go
Comment (tzneal) [inline pkg/features/kube_features.go:232]: Would this be 1.29 now?

**Review by tzneal — COMMENTED** [review]
Update the comment for DisableKubeletCloudCredentialProviders at https://github.com/kubernetes/kubernetes/blob/4619f7e9d9d64a0ae0b17270496ca3f7c67d24e8/pkg/features/kube_features.go#L240C2-L240C40 as well?

File: pkg/features/kube_features.go
Comment (dims) [reply pkg/features/kube_features.go:232]: Fixed. thanks!

**Comment by dims** [comment]
> Update the comment for DisableKubeletCloudCredentialProviders at https://github.com/kubernetes/kubernetes/blob/4619f7e9d9d64a0ae0b17270496ca3f7c67d24e8/pkg/features/kube_features.go#L240C2-L240C40 as well?

Done!

**Review by tzneal — APPROVED** [review]
/lgtm

**Comment by aojea** [comment]
Thanks dims

**Comment by aojea** [comment]
cc: @andrewsykim @liggitt

**Comment by aojea** [comment]
> @aojea please look at failures in the worker node kubelet in the latest `pull-kubernetes-e2e-gci-gce-ingress` CI job:
> 
> from log - https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/117503/pull-kubernetes-e2e-gci-gce-ingress/1697765618037559296/artifacts/bootstrap-e2e-minion-group-15cv/kubelet.log:
> 
> ```
> Sep 02 00:46:52.046603 bootstrap-e2e-minion-group-15cv kubelet[10550]: E0902 00:46:52.046567   10550 run.go:74] "command failed" err="failed to run Kubelet: cannot create certificate signing request: certificatesigningrequests.certificates.k8s.io is forbidden: User \"kubelet\" cannot create resource \"certificatesigningrequests\" in API group \"certificates.k8s.io\" at the cluster scope"
> ```

there are more failures in other components

```
role=extension-apiserver-authentication-reader --serviceaccount=YOUR_NS:YOUR_SA'
unable to load configmap based request-header-client-ca-file: configmaps "extension-apiserver-authentication" is forbidden: User "system:cloud-controller-manager" cannot get resource "configmaps" in API group "" in the namespace "kube-system"
Error: unable to load configmap based request-header-client-ca-file: configmaps "extension-apiserver-authentication" is forbidden: User "system:cloud-controller-manager" cannot get resource "configmaps" in API group "" in the namespace "kube-system"
Usage:
```

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/117503/pull-kubernetes-e2e-gci-gce-ingress/1697765618037559296/artifacts/bootstrap-e2e-master/cloud-controller-manager.log

it seems to be a more general rbac problem

**Comment by aojea** [comment]
the kube-addong-manager is not able to register

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/117503/pull-kubernetes-e2e-gci-gce-ingress/1697765618037559296/artifacts/bootstrap-e2e-master/kube-addon-manager.log

```
Error from server (NotFound): serviceaccounts "default" not found
WRN: == Error getting default service account, retry in 0.5 second ==
```

**Comment by aojea** [comment]
the kube-master-configuration  scripts fails
```
master","boot_id":"d7b51975c2f94a478a4346b249c1a8dd","timestamp":"2023-09-02T00:45:40.002119670Z","bootstrap_status":{"step_name":"UpdateLegacyAddonNodeLabels","status":"COMPLETED","status_reason":"","latency":"1.23726s"}}
Sep 02 00:45:48.570082 bootstrap-e2e-master configure-helper.sh[12179]: error: the server doesn't have a resource type "volumesnapshotclasses"
Sep 02 00:45:58.668565 bootstrap-e2e-master configure-helper.sh[12327]: error: the server doesn't have a resource type "volumesnapshotclasses"
Sep 02 00:46:08.861606 bootstrap-e2e-master configure-helper.sh[12482]: error: the server doesn't have a resource type "volumesnapshotclasses"
Sep 02 00:46:19.034558 bootstrap-e2e-master configure-helper.sh[12649]: error: the server doesn't have a resource type "volumesnapshotclasses"
Sep 02 00:46:29.131068 bootstrap-e2e-master configure-helper.sh[12797]: error: the server doesn't have a resource type "volumesnapshotclasses"
Sep 02 00:46:39.230542 bootstrap-e2e-master configure-helper.sh[13041]: error: the server doesn't have a resource type "volumesnapshotclasses"
``` 

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/117503/pull-kubernetes-e2e-gci-gce-ingress/1697765618037559296/artifacts/bootstrap-e2e-master/kube-master-configuration.log

File: cluster/gce/config-default.sh
Comment (aojea) [inline cluster/gce/config-default.sh:562]:
```
if [[ "${CLOUD_PROVIDER_FLAG:-}" == "external" ]]; then
  export ENABLE_AUTH_PROVIDER_GCP=true
```

**Comment by sftim** [comment]
Changelog suggestion
```diff
-In tree cloud providers are now switched off by default. Please use DisableCloudProviders and DisableKubeletCloudCredentialProvider feature flags if you still need this functionality.
+In-tree cloud provider integrations are now switched off by default. Please use the `DisableCloudProviders` and `DisableKubeletCloudCredentialProvider` feature flags if you still need this functionality.
```

**Comment by sftim** [comment]
Relevant to https://github.com/kubernetes/enhancements/issues/2395

**Pull request merged**
